### PR TITLE
C# reference: Consolidate exception-handling statements articles

### LINKED ISSUE #10749: Abridged Reference: Statements / exception statements
This will cover:

- `throw`
- `try` / `catch`
- `try` / `finally`
- `try` / `catch` / `finally`
- `when` (exception filter)

Sub-tasks are:

- [ ] Create the new articles
- [ ] Search for existing issues on content being migrated and consolidated. This is a good opportunity to fix those.
- [ ] Redirect existing content to the new articles.
- [ ] Remove old articles.
- [ ] Update all links in the .NET Content

### PR DESCRIPTION
**Pull request by pkulikov**
Fixes #10749

@BillWagner related to the exception doc content: #34888 (It also looks fine to redirect those tutorials to the new reference article)

---

#### Internal previews

<details><summary><strong>Toggle expand/collapse</strong></summary><br/>

| 📄 File | 🔗 Preview link |
|:--|:--|
| [docs/csharp/advanced-topics/expression-trees/index.md](https://github.com/dotnet/docs/blob/af1ca89fd830fbab67ae1f540345b6d077ce9d63/docs/csharp/advanced-topics/expression-trees/index.md) | [Expression Trees](https://review.learn.microsoft.com/en-us/dotnet/csharp/advanced-topics/expression-trees/index?branch=pr-en-us-35048) |
| [docs/csharp/asynchronous-programming/async-return-types.md](https://github.com/dotnet/docs/blob/af1ca89fd830fbab67ae1f540345b6d077ce9d63/docs/csharp/asynchronous-programming/async-return-types.md) | [Async return types (C#)](https://review.learn.microsoft.com/en-us/dotnet/csharp/asynchronous-programming/async-return-types?branch=pr-en-us-35048) |
| [docs/csharp/asynchronous-programming/cancel-async-tasks-after-a-period-of-time.md](https://github.com/dotnet/docs/blob/af1ca89fd830fbab67ae1f540345b6d077ce9d63/docs/csharp/asynchronous-programming/cancel-async-tasks-after-a-period-of-time.md) | [Cancel async tasks after a period of time](https://review.learn.microsoft.com/en-us/dotnet/csharp/asynchronous-programming/cancel-async-tasks-after-a-period-of-time?branch=pr-en-us-35048) |
| [docs/csharp/asynchronous-programming/task-asynchronous-programming-model.md](https://github.com/dotnet/docs/blob/af1ca89fd830fbab67ae1f540345b6d077ce9d63/docs/csharp/asynchronous-programming/task-asynchronous-programming-model.md) | [Task asynchronous programming model](https://review.learn.microsoft.com/en-us/dotnet/csharp/asynchronous-programming/task-asynchronous-programming-model?branch=pr-en-us-35048) |
| [docs/csharp/fundamentals/coding-style/coding-conventions.md](https://github.com/dotnet/docs/blob/af1ca89fd830fbab67ae1f540345b6d077ce9d63/docs/csharp/fundamentals/coding-style/coding-conventions.md) | [C# Coding Conventions](https://review.learn.microsoft.com/en-us/dotnet/csharp/fundamentals/coding-style/coding-conventions?branch=pr-en-us-35048) |
| [docs/csharp/fundamentals/exceptions/compiler-generated-exceptions.md](https://github.com/dotnet/docs/blob/af1ca89fd830fbab67ae1f540345b6d077ce9d63/docs/csharp/fundamentals/exceptions/compiler-generated-exceptions.md) | [Compiler-generated exceptions](https://review.learn.microsoft.com/en-us/dotnet/csharp/fundamentals/exceptions/compiler-generated-exceptions?branch=pr-en-us-35048) |
| [docs/csharp/fundamentals/exceptions/creating-and-throwing-exceptions.md](https://github.com/dotnet/docs/blob/af1ca89fd830fbab67ae1f540345b6d077ce9d63/docs/csharp/fundamentals/exceptions/creating-and-throwing-exceptions.md) | [Creating and Throwing Exceptions](https://review.learn.microsoft.com/en-us/dotnet/csharp/fundamentals/exceptions/creating-and-throwing-exceptions?branch=pr-en-us-35048) |
| [docs/csharp/fundamentals/exceptions/exception-handling.md](https://github.com/dotnet/docs/blob/af1ca89fd830fbab67ae1f540345b6d077ce9d63/docs/csharp/fundamentals/exceptions/exception-handling.md) | [Exception Handling (C# Programming Guide)](https://review.learn.microsoft.com/en-us/dotnet/csharp/fundamentals/exceptions/exception-handling?branch=pr-en-us-35048) |
| [docs/csharp/fundamentals/exceptions/how-to-execute-cleanup-code-using-finally.md](https://github.com/dotnet/docs/blob/af1ca89fd830fbab67ae1f540345b6d077ce9d63/docs/csharp/fundamentals/exceptions/how-to-execute-cleanup-code-using-finally.md) | [How to execute cleanup code using finally](https://review.learn.microsoft.com/en-us/dotnet/csharp/fundamentals/exceptions/how-to-execute-cleanup-code-using-finally?branch=pr-en-us-35048) |
| [docs/csharp/fundamentals/exceptions/how-to-handle-an-exception-using-try-catch.md](https://github.com/dotnet/docs/blob/af1ca89fd830fbab67ae1f540345b6d077ce9d63/docs/csharp/fundamentals/exceptions/how-to-handle-an-exception-using-try-catch.md) | [How to handle an exception using try/catch](https://review.learn.microsoft.com/en-us/dotnet/csharp/fundamentals/exceptions/how-to-handle-an-exception-using-try-catch?branch=pr-en-us-35048) |
| [docs/csharp/fundamentals/exceptions/index.md](https://github.com/dotnet/docs/blob/af1ca89fd830fbab67ae1f540345b6d077ce9d63/docs/csharp/fundamentals/exceptions/index.md) | [Exceptions and Exception Handling](https://review.learn.microsoft.com/en-us/dotnet/csharp/fundamentals/exceptions/index?branch=pr-en-us-35048) |
| [docs/csharp/fundamentals/tutorials/classes.md](https://github.com/dotnet/docs/blob/af1ca89fd830fbab67ae1f540345b6d077ce9d63/docs/csharp/fundamentals/tutorials/classes.md) | [Explore object oriented programming with classes and objects](https://review.learn.microsoft.com/en-us/dotnet/csharp/fundamentals/tutorials/classes?branch=pr-en-us-35048) |
| [docs/csharp/language-reference/compiler-messages/cs0163.md](https://github.com/dotnet/docs/blob/af1ca89fd830fbab67ae1f540345b6d077ce9d63/docs/csharp/language-reference/compiler-messages/cs0163.md) | [Compiler Error CS0163](https://review.learn.microsoft.com/en-us/dotnet/csharp/language-reference/compiler-messages/cs0163?branch=pr-en-us-35048) |
| [docs/csharp/language-reference/keywords/index.md](https://github.com/dotnet/docs/blob/af1ca89fd830fbab67ae1f540345b6d077ce9d63/docs/csharp/language-reference/keywords/index.md) | [C# Keywords](https://review.learn.microsoft.com/en-us/dotnet/csharp/language-reference/keywords/index?branch=pr-en-us-35048) |
| [docs/csharp/language-reference/keywords/statement-keywords.md](https://github.com/dotnet/docs/blob/af1ca89fd830fbab67ae1f540345b6d077ce9d63/docs/csharp/language-reference/keywords/statement-keywords.md) | [Statement keywords (C# Reference)](https://review.learn.microsoft.com/en-us/dotnet/csharp/language-reference/keywords/statement-keywords?branch=pr-en-us-35048) |
| [docs/csharp/language-reference/keywords/when.md](https://github.com/dotnet/docs/blob/af1ca89fd830fbab67ae1f540345b6d077ce9d63/docs/csharp/language-reference/keywords/when.md) | [when (C# reference)](https://review.learn.microsoft.com/en-us/dotnet/csharp/language-reference/keywords/when?branch=pr-en-us-35048) |
| [docs/csharp/language-reference/operators/await.md](https://github.com/dotnet/docs/blob/af1ca89fd830fbab67ae1f540345b6d077ce9d63/docs/csharp/language-reference/operators/await.md) | [await operator - asynchronously await for a task to complete](https://review.learn.microsoft.com/en-us/dotnet/csharp/language-reference/operators/await?branch=pr-en-us-35048) |
| [docs/csharp/language-reference/operators/null-coalescing-operator.md](https://github.com/dotnet/docs/blob/af1ca89fd830fbab67ae1f540345b6d077ce9d63/docs/csharp/language-reference/operators/null-coalescing-operator.md) | [?? and ??= operators - the null-coalescing operators](https://review.learn.microsoft.com/en-us/dotnet/csharp/language-reference/operators/null-coalescing-operator?branch=pr-en-us-35048) |
| [docs/csharp/language-reference/statements/exception-handling-statements.md](https://github.com/dotnet/docs/blob/af1ca89fd830fbab67ae1f540345b6d077ce9d63/docs/csharp/language-reference/statements/exception-handling-statements.md) | [docs/csharp/language-reference/statements/exception-handling-statements](https://review.learn.microsoft.com/en-us/dotnet/csharp/language-reference/statements/exception-handling-statements?branch=pr-en-us-35048) |
| [docs/csharp/language-reference/statements/jump-statements.md](https://github.com/dotnet/docs/blob/af1ca89fd830fbab67ae1f540345b6d077ce9d63/docs/csharp/language-reference/statements/jump-statements.md) | [Jump statements - `break`, `continue`, `return`, and `goto`](https://review.learn.microsoft.com/en-us/dotnet/csharp/language-reference/statements/jump-statements?branch=pr-en-us-35048) |
| [docs/csharp/language-reference/statements/lock.md](https://github.com/dotnet/docs/blob/af1ca89fd830fbab67ae1f540345b6d077ce9d63/docs/csharp/language-reference/statements/lock.md) | [lock statement - ensure exclusive access to a shared resource](https://review.learn.microsoft.com/en-us/dotnet/csharp/language-reference/statements/lock?branch=pr-en-us-35048) |
| [docs/csharp/language-reference/statements/selection-statements.md](https://github.com/dotnet/docs/blob/af1ca89fd830fbab67ae1f540345b6d077ce9d63/docs/csharp/language-reference/statements/selection-statements.md) | [Selection statements - `if`, `else`, and `switch`](https://review.learn.microsoft.com/en-us/dotnet/csharp/language-reference/statements/selection-statements?branch=pr-en-us-35048) |
| [docs/csharp/misc/cs0155.md](https://github.com/dotnet/docs/blob/af1ca89fd830fbab67ae1f540345b6d077ce9d63/docs/csharp/misc/cs0155.md) | [Compiler Error CS0155](https://review.learn.microsoft.com/en-us/dotnet/csharp/misc/cs0155?branch=pr-en-us-35048) |
| [docs/csharp/misc/cs0156.md](https://github.com/dotnet/docs/blob/af1ca89fd830fbab67ae1f540345b6d077ce9d63/docs/csharp/misc/cs0156.md) | [Compiler Error CS0156](https://review.learn.microsoft.com/en-us/dotnet/csharp/misc/cs0156?branch=pr-en-us-35048) |
| [docs/csharp/misc/cs0157.md](https://github.com/dotnet/docs/blob/af1ca89fd830fbab67ae1f540345b6d077ce9d63/docs/csharp/misc/cs0157.md) | [Compiler Error CS0157](https://review.learn.microsoft.com/en-us/dotnet/csharp/misc/cs0157?branch=pr-en-us-35048) |
| [docs/csharp/misc/cs0160.md](https://github.com/dotnet/docs/blob/af1ca89fd830fbab67ae1f540345b6d077ce9d63/docs/csharp/misc/cs0160.md) | [Compiler Error CS0160](https://review.learn.microsoft.com/en-us/dotnet/csharp/misc/cs0160?branch=pr-en-us-35048) |
| [docs/csharp/misc/cs0255.md](https://github.com/dotnet/docs/blob/af1ca89fd830fbab67ae1f540345b6d077ce9d63/docs/csharp/misc/cs0255.md) | [Compiler Error CS0255](https://review.learn.microsoft.com/en-us/dotnet/csharp/misc/cs0255?branch=pr-en-us-35048) |
| [docs/csharp/misc/cs1015.md](https://github.com/dotnet/docs/blob/af1ca89fd830fbab67ae1f540345b6d077ce9d63/docs/csharp/misc/cs1015.md) | [Compiler Error CS1015](https://review.learn.microsoft.com/en-us/dotnet/csharp/misc/cs1015?branch=pr-en-us-35048) |
| [docs/csharp/programming-guide/classes-and-structs/using-properties.md](https://github.com/dotnet/docs/blob/af1ca89fd830fbab67ae1f540345b6d077ce9d63/docs/csharp/programming-guide/classes-and-structs/using-properties.md) | [Using Properties (C# Programming Guide)](https://review.learn.microsoft.com/en-us/dotnet/csharp/programming-guide/classes-and-structs/using-properties?branch=pr-en-us-35048) |
| [docs/fundamentals/code-analysis/style-rules/ide0016.md](https://github.com/dotnet/docs/blob/af1ca89fd830fbab67ae1f540345b6d077ce9d63/docs/fundamentals/code-analysis/style-rules/ide0016.md) | [Use throw expression (IDE0016)](https://review.learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/style-rules/ide0016?branch=pr-en-us-35048) |

</details>

> **Note**
> This table shows preview links for the 30 files with the most changes. For preview links for other files in this PR, select <strong>OpenPublishing.Build Details</strong> within [checks](https://github.com/dotnet/docs/pull/35048/checks).

### CODE REVIEW

File: docs/csharp/toc.yml [hunk 1]
```diff
@@ -1134,6 +1124,9 @@ items:
     - name: Declaration statements
       displayName: var, ref
       href: language-reference/statements/declarations.md
+    - name: Exception-handling statements
+      displayName: throw, try, catch, finally, when, error, fault, try-catch, try-finally, try-catch-finally
```
Comment (BillWagner) [inline docs/csharp/toc.yml:1128]: The "try-catch", "try-finally", and "try-catch-finally" aren't needed. Those phrases match the "try", "catch" or "finally" keywords:

```suggestion
      displayName: throw, try, catch, finally, when, error, fault
```

File: docs/csharp/language-reference/statements/exception-handling-statements.md [hunk 2]
```diff
@@ -0,0 +1,141 @@
+---
+title: "Exception-handling statements - throw and try"
```
Comment (BillWagner) [inline docs/csharp/language-reference/statements/exception-handling-statements.md:2]: We should add catch here for SEO purposes:

```suggestion
title: "Exception-handling statements - throw, catch and try"
```

File: docs/csharp/language-reference/statements/exception-handling-statements.md [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+
+You use the `throw` and `try` statements to work with exceptions. Use the [`throw` statement](#the-throw-statement) to throw an exception. Use the [`try` statement](#the-try-statement) to catch and handle exceptions that might occur during execution of a code block.
+
+## The `throw` statement
+
+The `throw` statement throws an exception:
+
+:::code language="csharp" source="snippets/exception-handling-statements/Program.cs" id="Throw":::
+
+In a `throw e;` statement, the result of expression `e` must be implicitly convertible to <xref:System.Exception?displayProperty=nameWithType>. You can use the built-in exception classes, for example, <xref:System.ArgumentOutOfRangeException> or <xref:System.InvalidOperationException>. You can also define your own exception classes that derive from <xref:System.Exception?displayProperty=nameWithType>. For more information, see [Creating and throwing exceptions](../../fundamentals/exceptions/creating-and-throwing-exceptions.md).
```
Comment (BillWagner) [inline docs/csharp/language-reference/statements/exception-handling-statements.md:34]: What do you think about mentioning helpers like "ThrowIfNull" here:

```suggestion
In a `throw e;` statement, the result of expression `e` must be implicitly convertible to <xref:System.Exception?displayProperty=nameWithType>. You can use the built-in exception classes, for example, <xref:System.ArgumentOutOfRangeException> or <xref:System.InvalidOperationException>. You can also use helper methods like <xref:System.ArgumentNullException.ThrowIfNull%2A?displayProperty=nameWithType> which throw exceptions based on conditions. You can also define your own exception classes that derive from <xref:System.Exception?displayProperty=nameWithType>. For more information, see [Creating and throwing exceptions](../../fundamentals/exceptions/creating-and-throwing-exceptions.md).
```

File: docs/csharp/language-reference/statements/exception-handling-statements.md [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+When an exception occurs, catch clauses are examined in the specified order, from top to bottom. At maximum, only one `catch` block is executed for any thrown exception. As the preceding example also shows, you can omit declaration of an exception variable and specify only the exception type in a catch clause. A catch clause without any specified exception type matches any exception and, if present, must be the last catch clause.
+
+If you want to re-throw a caught exception, use the [`throw` statement](#the-throw-statement), as the following example shows:
+
+:::code language="csharp" source="snippets/exception-handling-statements/Program.cs" id="RethrowInCatch":::
+
+> [!NOTE]
+> `throw;` preserves the original stack trace of the exception, which is stored in the <xref:System.Exception.StackTrace?displayProperty=nameWithType> property. Opposite to that, `throw e;` updates the <xref:System.Exception.StackTrace> property of `e`.
+
+#### A `when` exception filter
```
Comment (BillWagner) [inline docs/csharp/language-reference/statements/exception-handling-statements.md:84]: We need to add text here to explain that exception filters modify the rules on the order of `catch` clauses.  Namely:

1. Multiple `catch` clauses can specify the same `Exception` type, if `when` clauses distinguish them.
2. The `catch Exception` clause doesn't need to be last *if* an exception filter is present.

File: docs/csharp/language-reference/statements/exception-handling-statements.md [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+
+Along with an exception type, you can also specify an exception filter that further examines an exception and decides if the corresponding `catch` block handles that exception. An exception filter is a Boolean expression that follows the `when` keyword, as the following example shows:
+
+:::code language="csharp" source="snippets/exception-handling-statements/Program.cs" id="WhenFilter":::
+
+The preceding example uses an exception filter to provide a single `catch` block to handle exceptions of two specified types.
+
+#### Exceptions in async methods
```
Comment (BillWagner) [inline docs/csharp/language-reference/statements/exception-handling-statements.md:92]: I'd add a paragraph that similar behavior happens in iterator methods: The exception is thrown when the iterator is advanced using `MoveNext`

**Review by BillWagner — COMMENTED** [review]
Overall, this is great @pkulikov 

I had a few comments, most of which probably applied before these edits. Once those are made, this is ready.

File: docs/csharp/language-reference/statements/exception-handling-statements.md (hunk 3 above)
Comment (pkulikov) [reply docs/csharp/language-reference/statements/exception-handling-statements.md:34]: I like that idea, thanks!

**Comment by pkulikov** [comment]
@BillWagner thank you for the review. I've addressed all the feedback, please check the latest commit.

**Review by BillWagner — APPROVED** [review]
Thanks again @pkulikov 

This latest version looks great. I'll :shipit:

**Pull request merged**